### ISSUE TITLE
-Wunused-variable triggers for temporaries with side effects when bound to a reference

### ISSUE THREAD

**Issue by cor3ntin** [llvm/llvm-project#54489]
The following code warns on `guard2` and `guard3` (but not guard) despite all 3 variable having side effects on destruction.
This feels inconsistent and surprising.
GCC emits no warning in any of these cases.

```cpp

struct RAIIWrapper {
    RAIIWrapper();
    ~RAIIWrapper();
};

void foo() {
    auto const  guard = RAIIWrapper();
    auto const& guard2 = RAIIWrapper(); //unused variable 'guard2'
    auto && guard3 = RAIIWrapper(); //unused variable 'guard3'
}
```

https://godbolt.org/z/zr7dsevoc

**Comment by dwblaikie** [comment]
My personal hope/take on this is that the warning should be aggressive like this (because it lets us warn on things like unused `std::string`), but we should have an attribute for types with scoped side effects (and then code like `RAIIWrapper(), func();` is valid - since `func()` happens between the wrapper's ctor and dtor, but `RAIIWrapper();` is still "unused" - but there might be a third type with arbitrary (not even scoped) side effects, where `type_name();` would be valid/not-warned)

**Comment by dwblaikie** [comment]
Oh, yeah, I guess this is just about reference lifetime extended variables - this doesn't lead us to being able to warn on std::string currently.

Yeah, we shouldn't warn on the reference extended cases anymore than we should on the direct variable cases.
